File: backend/methods/bisection.py
from .utils import evaluate_function

def bisection_method(func_str, a, b, tol=1e-6, max_iter=100):
    """
    Solves for the root of a function using the Bisection Method.
    Returns the root, a message, and a list of step dictionaries for the iteration table.
    """
    steps = []
    
    fa = evaluate_function(func_str, a)
    fb = evaluate_function(func_str, b)
    
    if fa * fb >= 0:
        return {
            "success": False,
            "message": "Function must have opposite signs at a and b (f(a) * f(b) < 0).",
            "root": None,
            "steps": steps
        }
    
    for i in range(1, max_iter + 1):
        midpoint = (a + b) / 2.0
        f_mid = evaluate_function(func_str, midpoint)
        error = abs(b - a) / 2.0
        
        step_data = {
            "iteration": i,
            "a": a,
            "b": b,
            "midpoint": midpoint,
            "f_mid": f_mid,
            "error": error
        }
        steps.append(step_data)
        
        if error < tol or f_mid == 0:
            return {
                "success": True,
                "message": f"Converged to root after {i} iterations.",
                "root": midpoint,
                "steps": steps
            }
            
        if fa * f_mid < 0:
            b = midpoint
            fb = f_mid
        else:
            a = midpoint
            fa = f_mid
            
    return {
        "success": False,
        "message": f"Did not converge within {max_iter} iterations.",
        "root": midpoint,
        "steps": steps
    }

File: worker/src/methods/bisection.ts
import { evaluateFunction } from "../lib/exprEval";

export interface BisectionStep {
  iteration: number;
  a: number;
  b: number;
  midpoint: number;
  f_mid: number;
  error: number;
}

export function bisectionMethod(
  funcStr: string,
  a: number,
  b: number,
  tol = 1e-6,
  maxIter = 100
) {
  const steps: BisectionStep[] = [];

  const fa0 = evaluateFunction(funcStr, a);
  const fb0 = evaluateFunction(funcStr, b);

  if (fa0 * fb0 >= 0) {
    return {
      success: false,
      message: "Function must have opposite signs at a and b (f(a) * f(b) < 0).",
      root: null,
      steps,
    };
  }

  let fa = fa0;
  let midpoint = (a + b) / 2.0;

  for (let i = 1; i <= maxIter; i++) {
    midpoint = (a + b) / 2.0;
    const fMid = evaluateFunction(funcStr, midpoint);
    const error = Math.abs(b - a) / 2.0;

    steps.push({ iteration: i, a, b, midpoint, f_mid: fMid, error });

    if (error < tol || fMid === 0) {
      return {
        success: true,
        message: `Converged to root after ${i} iterations.`,
        root: midpoint,
        steps,
      };
    }

    if (fa * fMid < 0) {
      b = midpoint;
    } else {
      a = midpoint;
      fa = fMid;
    }
  }

  return {
    success: false,
    message: `Did not converge within ${maxIter} iterations.`,
    root: midpoint,
    steps,
  };
}
